File: incre-tests/synduce/constraints/balanced_tree/node_count.f
Inductive Tree = nil Unit | node {Int, Tree, Tree};

max = \a: Int. \b: Int. if < a b then b else a;

height = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    nil _ -> 0
  | node {w, l, r} -> + 1 (max (f l) (f r))
  end
);

balanced = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    nil _ -> true
  | node {w, l, r} -> and (and (== (height l) (height r)) (f l)) (f r)
  end
);

spec = fix (
  \f : Tree -> Int. \t: Tree.
  match t with
    nil _ -> 0
  | node {w, l, r} -> + 1 (+ (f l) (f r))
  end
);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | node {w, l, r} -> node {w, f l, r} /*Avoid the recursion of r*/
  end
);

main = \t: Tree. if balanced t then spec (target t) else 0;

File: ml-benchmark/synduce/constraints/balanced_tree/node_count.ml
type tree = Nil of unit | Node of int * tree * tree

let max a b = if a < b then b else a

let rec height t =
  match t with
  | Nil _ -> 0
  | Node (w, l, r) -> 1 + max (height l) (height r)

let rec balanced t =
  match t with
  | Nil _ -> true
  | Node (w, l, r) -> (height l = height r) && balanced l && balanced r

let rec spec t =
  match t with
  | Nil _ -> 0
  | Node (w, l, r) -> 1 + (spec l + spec r)

let rec target t =
  match t with
  | Nil _ -> Nil ()
  | Node (w, l, r) -> Node (w, target l, r)

let program t =
  if balanced t then spec (target t) else 0